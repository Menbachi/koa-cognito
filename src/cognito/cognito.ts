import jwt from 'jsonwebtoken';
import jwkToPem, { JWK } from 'jwk-to-pem';

export interface CognitoConfiguration {
  region: string;
  userPoolId: string;
}

export const compareTokenWithKid = () => {};
export const verifyTheClaims = () => {};

export type Pems = { [key: string]: string };

export const getPems = async ({ region, userPoolId }: CognitoConfiguration): Promise<Pems> => {
  const url = `https://cognito-idp.${region}.amazonaws.com/${userPoolId}/.well-known/jwks.json`;
  const response = await fetch(url);
  const jwk = await response.json();

  // @TODO - find better way for typesafe kid & JWK
  return jwk?.keys?.reduce((prev: any, current: { kid: any } & jwkToPem.JWK) => {
    return { ...prev, [current.kid]: jwkToPem(current) };
  }, {});
};

export const getAllPems = async (configurations: CognitoConfiguration[]): Promise<Pems> => {
  const allPems = await Promise.all(configurations.map((config) => getPems(config)));

  return allPems.reduce((prev, current) => ({ ...prev, ...current }), {});
};

export const testToken = async (configuration: CognitoConfiguration[]) => {
  const pems = await getAllPems(configuration);

  return async (token: string) => {
    const decodedToken = jwt.decode(token, { complete: true });
    if (!decodedToken) {
      console.warn(`Cannot decode token: ${token}`);
      return undefined;
    }

    if (!Object.values(pems).some((pemIssuer) => decodedToken.payload.iss === pemIssuer)) {
      console.warn(`Cannod find pem issuer ${decodedToken.payload.iss}, pems: ${pems}`);
      return undefined;
    }
    const kid = decodedToken.header.kid
    if(!kid) {
      console.warn(`Cannot find kid in token`);
      return undefined;
    }
    const pem = pems[kid];
    if (!pem) {
      console.warn(`Cannot find pem by kid: ${kid}, pems: ${pems}`);
      return undefined;
    }

    return jwt.verify(token, pem, { issuer: decodedToken.payload.iss })
  };
};
