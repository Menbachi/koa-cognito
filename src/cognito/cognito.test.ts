import { getAllPems, getPems } from './cognito';

const createKey = (kid: string) => ({
  alg: 'RS256',
  e: 'AQAB',
  kid,
  kty: 'RSA',
  n: 'lsjhglskjhgslkjgh43lj5h34lkjh34lkjht3example',
  use: 'sig',
});

describe('Pems', () => {
  test('Response with keys for pems', async () => {
    global.fetch = jest.fn().mockResolvedValue({
      json: () => ({
        keys: [createKey('kid-A'), createKey('kid-B'), createKey('kid-C')],
      }),
    });
    const configuration = { region: 'exampleRegion', userPoolId: 'exampleUserPoolId' };
    const response = await getPems(configuration);

    expect(global.fetch).toBeCalledWith(
      `https://cognito-idp.${configuration.region}.amazonaws.com/${configuration.userPoolId}/.well-known/jwks.json`
    );
    expect(response).toMatchSnapshot();
  });

  test('Response with keys for pems for all configuration regions', async () => {
    global.fetch = jest
      .fn()
      .mockResolvedValueOnce({
        json: () => ({
          keys: [createKey('kid-A')],
        }),
      })
      .mockResolvedValueOnce({
        json: () => ({
          keys: [createKey('kid-C')],
        }),
      });
    const configuration = [
      { region: 'exampleRegion1', userPoolId: 'exampleUserPoolId1' },
      { region: 'exampleRegion2', userPoolId: 'exampleUserPoolId2' },
    ];
    const response = await getAllPems(configuration);

    expect(global.fetch).toHaveBeenCalledTimes(2);
    expect(global.fetch).toHaveBeenCalledWith(
      `https://cognito-idp.${configuration[0].region}.amazonaws.com/${configuration[0].userPoolId}/.well-known/jwks.json`
    );
    expect(global.fetch).toHaveBeenCalledWith(
      `https://cognito-idp.${configuration[1].region}.amazonaws.com/${configuration[1].userPoolId}/.well-known/jwks.json`
    );
    console.log({ response });
    expect(response).toMatchSnapshot();
  });
});
