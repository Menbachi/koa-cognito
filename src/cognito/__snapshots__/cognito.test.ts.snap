// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pems Response with keys for pems for all configuration regions 1`] = `
Object {
  "0": Object {
    "kid-A": "-----BEGIN PUBLIC KEY-----
MD0wDQYJKoZIhvcNAQEBBQADLAAwKQIiAJbI4YJbJI4YLJZI4IeN5Y+Yd+JZI4d+
JZI4bd3sWpqZXgIDAQAB
-----END PUBLIC KEY-----
",
  },
  "1": Object {
    "kid-C": "-----BEGIN PUBLIC KEY-----
MD0wDQYJKoZIhvcNAQEBBQADLAAwKQIiAJbI4YJbJI4YLJZI4IeN5Y+Yd+JZI4d+
JZI4bd3sWpqZXgIDAQAB
-----END PUBLIC KEY-----
",
  },
}
`;
